src/lib.rs

```diff
@@ -79,16 +94,28 @@ fn longest_path(graph: &digraph::PyDiGraph) -> PyResult<Vec<usize>> {
         }
     };
     if nodes.is_empty() {
-        return Ok(path);
+        return Ok((path, 0));
     }
     let mut dist: HashMap<NodeIndex, (usize, NodeIndex)> = HashMap::new();
     for node in nodes {
-        let parents =
-            dag.neighbors_directed(node, petgraph::Direction::Incoming);
+        let parents = dag.edges_directed(node, petgraph::Direction::Incoming);
         let mut us: Vec<(usize, NodeIndex)> = Vec::new();
-        for p_node in parents {
-            let length = dist[&p_node].0 + 1;
-            us.push((length, p_node));
+        let mut local_dist: HashMap<NodeIndex, usize> = HashMap::new();
+        for p_edge in parents {
+            let p_node = p_edge.source();
+            let weight: usize = edge_weight_callable(
+                p_node.index(),
+                p_edge.target().index(),
+                p_edge.weight(),
+            )?;
+            local_dist
+                .entry(p_node)
+                .and_modify(|x| *x = cmp::max(*x, weight))
+                .or_insert(weight);
+        }
+        for (p_node, weight) in local_dist.iter() {
+            let length = dist[p_node].0 + weight;
+            us.push((length, *p_node));
```

Comment:
Is it really necessary to store the "local" weights? Can't we just do something like:
```suggestion
        for p_edge in parents {
            let p_node = p_edge.source();
            let weight: usize = edge_weight_callable(
                p_node.index(),
                p_edge.target().index(),
                p_edge.weight(),
            )?;
            let length = dist[&p_node].0 + weight;
            us.push((length, p_node));
```
since after all we only keep the maximum over `us`.

Reply:
My concern with doing this was in the parallel edge case won't `dist[&p_node].0` contain a sum of parallel edges from `p_node` to `node` while we want to have just the maximum weight.

Reply:
Oh, yep I was misreading the code (jumping back and forth between python and rust too much), yeah this is fine we're not modifying things in dist until we find the max.